### PR TITLE
[vcpkg baseline][qt5-base] Fix glib link issues on linux

### LINKED ISSUE #16942: [cmake|linux] build failure
see https://dev.azure.com/vcpkg/public/_build/results?buildId=51173&view=logs&jobId=e43d08a3-46a6-5c05-c22b-b8b3a7ac1517&j=e43d08a3-46a6-5c05-c22b-b8b3a7ac1517&t=341b2070-72fd-5058-0746-f056592aaa91
probably reason:

`harfbuzz` build without `glib` feature.
`qt5-base` build with `glib` being available <- thinks harfbuzz was build with glib enabled and links it for harfbuzz (or even links it itself; missing -no-glib flag?)
-> build error 

Needed to fix: Intropspection into already build ports and the selection of features those ports are build with!

### PR DESCRIPTION
**Pull request by Neumann-A**
Fix CI issues in:
#16948 
#16897 

closes #16942

Edit: Failure in above PRs:
c++: error: /mnt/vcpkg-ci/installed/x64-linux/debug/lib//libglib-2.0.a: No such file or directory
c++: error: /mnt/vcpkg-ci/installed/x64-linux/debug/lib//libglib-2.0.a: No such file or directory
c++: error: /mnt/vcpkg-ci/installed/x64-linux/debug/lib//libglib-2.0.a: No such file or directory

### CODE REVIEW

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
Ping @ras0219 @ras0219-msft for review this PR.
Can we have other better ways to achieve this?

**Comment by Neumann-A** [comment]
Just a note: Passing `--triplet=${TARGET_TRIPLET}` to the `list` command does nothing. (I would at least assume it will filter the list.)

File: ports/qt5-base/portfile.cmake [hunk 1]
```diff
@@ -1,5 +1,30 @@
...
+        message(FATAL_ERROR "x_vcpkg_get_port_features requires parameter PORTS!")
+    endif()
+    find_program(VCPKG NAMES vcpkg PATHS "${VCPKG_ROOT_DIR}")
+    message(STATUS "vcpkg: ${VCPKG}")
+    execute_process(COMMAND "${VCPKG}" list
+                                   WORKING_DIRECTORY "${VCPKG_ROOT_DIR}"
+                                   OUTPUT_VARIABLE _all_installed_ports
+                                   ERROR_VARIABLE _all_installed_ports_err)
+    string(REPLACE ";" "(\\\[[^:]+\\\])?|" _port_match_regex "${x_vcpkg_get_port_features_PORTS}")
+    set(_regex "(${_port_match_regex}(\\\[[^:]\\\])):${TARGET_TRIPLET}")
```
Comment (Neumann-A) [inline ports/qt5-base/portfile.cmake:15]:
```suggestion
```

File: ports/qt5-base/portfile.cmake [hunk 2]
```diff
@@ -1,5 +1,30 @@
...
+    cmake_parse_arguments(PARSE_ARGV 0 x_vcpkg_get_port_features "" "" "PORTS")
+    if(NOT x_vcpkg_get_port_features_PORTS)
+        message(FATAL_ERROR "x_vcpkg_get_port_features requires parameter PORTS!")
+    endif()
+    find_program(VCPKG NAMES vcpkg PATHS "${VCPKG_ROOT_DIR}")
+    message(STATUS "vcpkg: ${VCPKG}")
+    execute_process(COMMAND "${VCPKG}" list
+                                   WORKING_DIRECTORY "${VCPKG_ROOT_DIR}"
+                                   OUTPUT_VARIABLE _all_installed_ports
+                                   ERROR_VARIABLE _all_installed_ports_err)
```
Comment (Neumann-A) [inline ports/qt5-base/portfile.cmake:13]:
```suggestion
                                   )
```

**Review by vicroms — APPROVED** [review]
LGTM

**Comment by vicroms** [comment]
An alternative is to run `vcpkg x-package-info {port}:{triplet} --x-installed --x-json`, which will produce a `features` array as part of the JSON output.

File: ports/qt5-base/portfile.cmake [hunk 3]
```diff
@@ -222,12 +247,11 @@ set(FREETYPE_RELEASE_ALL "${FREETYPE_RELEASE} ${BZ2_RELEASE} ${LIBPNG_RELEASE} $
...
 
 # If HarfBuzz is built with GLib enabled, it must be statically link
-set(GLIB_LIB_VERSION 2.0)
-find_library(GLIB_RELEASE NAMES glib-${GLIB_LIB_VERSION} PATHS "${CURRENT_INSTALLED_DIR}/lib" NO_DEFAULT_PATH)
-find_library(GLIB_DEBUG NAMES glib-${GLIB_LIB_VERSION} PATHS "${CURRENT_INSTALLED_DIR}/debug/lib" NO_DEFAULT_PATH)
-if(GLIB_RELEASE MATCHES "-NOTFOUND" OR GLIB_DEBUG MATCHES "-NOTFOUND")
-    set(GLIB_RELEASE "")
-    set(GLIB_DEBUG "")
+x_vcpkg_get_port_features(PORTS harfbuzz)
+if("glib" IN_LIST harfbuzz_FEATURES)
```
Comment (ras0219-msft) [inline ports/qt5-base/portfile.cmake:251]: Should this be instead handled with `harfbuzz.pc`?

File: ports/qt5-base/portfile.cmake [hunk 4]
```diff
@@ -1,5 +1,29 @@
 vcpkg_buildpath_length_warning(37)
 
+function(x_vcpkg_get_port_features)
+    cmake_parse_arguments(PARSE_ARGV 0 x_vcpkg_get_port_features "" "" "PORTS")
+    if(NOT x_vcpkg_get_port_features_PORTS)
+        message(FATAL_ERROR "x_vcpkg_get_port_features requires parameter PORTS!")
+    endif()
+    find_program(VCPKG NAMES vcpkg PATHS "${VCPKG_ROOT_DIR}")
+    message(STATUS "vcpkg: ${VCPKG}")
+    execute_process(COMMAND "${VCPKG}" list
```
Comment (ras0219-msft) [inline ports/qt5-base/portfile.cmake:10]: There are a lot of potential problems with recursive vcpkg like this; we should find another way.

File: ports/qt5-base/portfile.cmake [hunk 5]
```diff
@@ -222,12 +247,11 @@ set(FREETYPE_RELEASE_ALL "${FREETYPE_RELEASE} ${BZ2_RELEASE} ${LIBPNG_RELEASE} $
 set(FREETYPE_DEBUG_ALL "${FREETYPE_DEBUG} ${BZ2_DEBUG} ${LIBPNG_DEBUG} ${ZLIB_DEBUG} ${BROTLI_DEC_DEBUG} ${BROTLI_COMMON_DEBUG}")
 
 # If HarfBuzz is built with GLib enabled, it must be statically link
-set(GLIB_LIB_VERSION 2.0)
-find_library(GLIB_RELEASE NAMES glib-${GLIB_LIB_VERSION} PATHS "${CURRENT_INSTALLED_DIR}/lib" NO_DEFAULT_PATH)
-find_library(GLIB_DEBUG NAMES glib-${GLIB_LIB_VERSION} PATHS "${CURRENT_INSTALLED_DIR}/debug/lib" NO_DEFAULT_PATH)
-if(GLIB_RELEASE MATCHES "-NOTFOUND" OR GLIB_DEBUG MATCHES "-NOTFOUND")
```
Comment (ras0219-msft) [inline ports/qt5-base/portfile.cmake:228]: While this approach is a bit ugly, is there a problem with "spuriously" including glib if it is available?

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
I'd really prefer to avoid this approach (recursive vcpkg to inspect installed files).

If nothing else, `harfbuzz` should install a file that we can test for existence (such as `share/harfbuzz/has_glib_feature`.

File: ports/qt5-base/portfile.cmake (hunk 5 above)
Comment (Neumann-A) [reply ports/qt5-base/portfile.cmake:228]: > "spuriously" including glib if it is available?
 
 Yes. CI failures due to expecting to have glib libraries available while they are not restored in CI.

File: ports/qt5-base/portfile.cmake (hunk 4 above)
Comment (Neumann-A) [reply ports/qt5-base/portfile.cmake:10]: ok, will change

File: ports/qt5-base/portfile.cmake [hunk 6]
```diff
@@ -1,20 +1,11 @@
...
+    set(QT_OPENSSL_LINK OFF)
 endif()
 
-if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
-    option(QT_MYSQL_PLUGIN "Create MySQL plugin" OFF)
-else()
-    if ("mysqlplugin" IN_LIST FEATURES)
-        option(QT_MYSQL_PLUGIN "Create MySQL plugin" ON)
-    endif()
-endif()
```
Comment (Neumann-A) [inline ports/qt5-base/portfile.cmake:15]: Who came up with this nonsense? Was probably a hack to make qt build on linux/osx due to the feature being enabled?

File: scripts/cmake/x_vcpkg_get_pkgconfig_modules.cmake [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    endif()
+
+    set(_prefix "${x_vcpkg_get_pkgconfig_modules_PREFIX}")
+    set(_modules "${x_vcpkg_get_pkgconfig_modules_MODULES}")
+
+    vcpkg_find_acquire_program(PKGCONFIG)
+    set(backup_PKG_CONFIG_PATH "$ENV{PKG_CONFIG_PATH}")
+
+    set(var_suffixes)
+    if(EXISTS "${CURRENT_INSTALLED_DIR}/lib/pkgconfig")
```
Comment (Neumann-A) [inline scripts/cmake/x_vcpkg_get_pkgconfig_modules.cmake:58]:
```suggestion
    if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")
```

File: scripts/cmake/x_vcpkg_get_pkgconfig_modules.cmake [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        if(x_vcpkg_get_pkgconfig_modules_INCLUDE_DIRS)
+            execute_process(
+                COMMAND "${PKGCONFIG}" --cflags-only-I ${_modules}
+                OUTPUT_VARIABLE ${_prefix}_INCLUDE_DIRS
+                OUTPUT_STRIP_TRAILING_WHITESPACE
+            )
+            list(APPEND var_suffixes INCLUDE_DIRS)
+        endif()
+    endif()
+    if(EXISTS "${CURRENT_INSTALLED_DIR}/debug/lib/pkgconfig")
```
Comment (Neumann-A) [inline scripts/cmake/x_vcpkg_get_pkgconfig_modules.cmake:93]:
```suggestion
    if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
```

File: scripts/cmake/x_vcpkg_get_port_info.cmake [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+            include("${_port_info}")
+        endif()
+        set(${_port}_FEATURES "${${_port}_FEATURES}" PARENT_SCOPE)
+    endforeach()
+endfunction()
+
+# Write information about the port. Keep in sync with x_vcpkg_get_port_info
+function(z_vcpkg_write_port_info)
+    set(_file "${CURRENT_PACKAGES_DIR}/share/${PORT}/vcpkg_port_info.cmake")
+    set(_contents "set(${PORT}_FEATURES \"${FEATURES}\")\n")
```
Comment (Neumann-A) [inline scripts/cmake/x_vcpkg_get_port_info.cmake:43]: could probably also use 
 `set(_contents "set(${PORT}_LINKAGE \"${VCPKG_LIBRARY_LINKAGE}\")\n")`
but i probably want to move this to a separate PR instead of waiting a day before this is mergeable

**Comment by PhoebeHui** [comment]
The forest issue would be fixed by PR https://github.com/microsoft/vcpkg/pull/16836

File: ports/qt5-base/portfile.cmake [hunk 10]
```diff
@@ -1,20 +1,11 @@
 vcpkg_buildpath_length_warning(37)
 
 if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
-    option(QT_OPENSSL_LINK "Link against OpenSSL at compile-time." ON)
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply ports/qt5-base/portfile.cmake:4]: I think that's possibly the objective.

File: docs/maintainers/x_vcpkg_get_port_info.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
+# x_vcpkg_get_port_info
```
Comment (ras0219-msft) [inline docs/maintainers/x_vcpkg_get_port_info.md:1]: I'd prefer to not add this now

File: scripts/cmake/x_vcpkg_get_pkgconfig_modules.cmake [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+Returns `"${PKGCONFIG}" --libs` in <prefix>_LIBS_(DEBUG|RELEASE)
+
+### LIBRARIES
+Returns `"${PKGCONFIG}" --libs-only-l` in <prefix>_LIBRARIES_(DEBUG|RELEASE)
+
+### LIBRARIES_DIRS
+Returns `"${PKGCONFIG}" --libs-only-L` in <prefix>_LIBRARIES_DIRS_(DEBUG|RELEASE)
+
+### INCLUDE_DIRS
+Returns `"${PKGCONFIG}"  --cflags-only-I` in <prefix>_INCLUDE_DIRS
```
Comment (ras0219-msft) [inline scripts/cmake/x_vcpkg_get_pkgconfig_modules.cmake:36]: Does this need to be split for DEBUG|RELEASE?

It's conceivable that a library could do something strange here.

File: docs/maintainers/x_vcpkg_get_port_info.md (hunk 11 above)
Comment (Neumann-A) [reply docs/maintainers/x_vcpkg_get_port_info.md:1]: ok but this will trigger a whole rebuild again

File: scripts/cmake/x_vcpkg_get_pkgconfig_modules.cmake (hunk 12 above)
Comment (Neumann-A) [reply scripts/cmake/x_vcpkg_get_pkgconfig_modules.cmake]: yeah i actually already saw something like this. glib installing a single config header into `/lib/include/glib/glib-2.0`

File: scripts/ports.cmake [hunk 13]
```diff
@@ -136,12 +136,15 @@ if(CMD MATCHES "^BUILD$")
...
 
+    include("${SCRIPTS}/cmake/x_vcpkg_get_pkgconfig_modules.cmake")
+
     include("${SCRIPTS}/cmake/z_vcpkg_apply_patches.cmake")
     include("${SCRIPTS}/cmake/z_vcpkg_prettify_command_line.cmake")
 
     include("${CURRENT_PORT_DIR}/portfile.cmake")
     if(DEFINED PORT)
         include("${SCRIPTS}/build_info.cmake")
+        z_vcpkg_write_port_info()
```
Comment (Neumann-A) [inline scripts/ports.cmake:147]:
```suggestion
```

**Comment by JackBoosY** [comment]
Ping @ras0219 @ras0219-msft for review this PR again.

File: ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+* [qt5-base](https://github.com/microsoft/vcpkg/blob/master/ports/qt5-base/portfile.cmake)
+#]===]
+if(Z_VCPKG_PKGCONFIG_GET_MODULES_GUARD)
+    return()
+endif()
+set(Z_VCPKG_PKGCONFIG_GET_MODULES_GUARD ON CACHE INTERNAL "guard variable")
+
+function(x_vcpkg_pkgconfig_get_modules)
+    cmake_parse_arguments(PARSE_ARGV 0 "arg" "LIBS;LIBRARIES;LIBRARIES_DIR;INCLUDE_DIRS;" "PREFIX" "MODULES")
```
Comment (strega-nil) [inline ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake:48]:
```suggestion
    cmake_parse_arguments(PARSE_ARGV 0 "arg" "LIBS;LIBRARIES;LIBRARIES_DIR;INCLUDE_DIRS" "PREFIX" "MODULES")
```

File: ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+    if(NOT arg_PREFIX)
```
Comment (strega-nil) [inline ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake:49]:
```suggestion
    if(NOT DEFINED arg_PREFIX)
```

File: ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        message(FATAL_ERROR "x_vcpkg_pkgconfig_get_modules requires parameter PREFIX!")
+    endif()
+    if(NOT arg_MODULES)
```
Comment (strega-nil) [inline ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake:52]:
```suggestion
    if(NOT DEFINED arg_MODULES)
```

File: ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+        message(FATAL_ERROR "x_vcpkg_pkgconfig_get_modules requires parameter MODULES!")
+    endif()
```
Comment (strega-nil) [inline ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake:54]:
```suggestion
    endif()
    if(DEFINED arg_UNPARSED_ARGUMENTS)
        message(FATAL_ERROR "extra arguments passed to x_vcpkg_pkgconfig_get_modules: ${arg_UNPARSED_ARGUMENTS}")
    endif()
```

File: ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+    set(ENV{PKG_CONFIG_PATH} "${backup_PKG_CONFIG_PATH}")
+
+    foreach(_var IN LISTS var_suffixes)
+        set("${arg_PREFIX}_${_var}" "${${arg_PREFIX}_${_var}}" PARENT_SCOPE)
+    endforeach()
+endfunction()
+
+function(z_vcpkg_set_pkgconfig_path _path _backup)
+    if(_backup)
```
Comment (strega-nil) [inline ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake:138]:
```suggestion
    if(NOT _backup STREQUAL "")
```

File: ports/vcpkg-pkgconfig-get-modules/portfile.cmake [hunk 19]
```diff
@@ -0,0 +1,12 @@
+if(NOT TARGET_TRIPLET STREQUAL _HOST_TRIPLET)
```
Comment (strega-nil) [inline ports/vcpkg-pkgconfig-get-modules/portfile.cmake:1]:
```suggestion
if(NOT TARGET_TRIPLET STREQUAL HOST_TRIPLET)
```

**Review by strega-nil — CHANGES_REQUESTED** [review]
minor cmake changes

File: ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake (hunk 15 above)
Comment (Neumann-A) [reply ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake:49]: Prefix should not be `""` although it shouldn't create any bigger issues.

File: ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake (hunk 16 above)
Comment (Neumann-A) [reply ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake:52]: arg_MODULES is not allowed to be `""`- Personally I think the error message is ok also in the case that somebody passed ""

File: ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake (hunk 15 above)
Comment (Neumann-A) [inline ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake:49]:
```suggestion
    if(NOT DEFINED arg_PREFIX OR arg_PREFIX STREQUAL "")

```

File: ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake (hunk 16 above)
Comment (Neumann-A) [inline ports/vcpkg-pkgconfig-get-modules/x_vcpkg_pkgconfig_get_modules.cmake:52]:
```suggestion
    if(NOT DEFINED arg_MODULES OR arg_MODULES STREQUAL "")
```

**Review by strega-nil — APPROVED** [review]

**Comment by vicroms** [comment]
Waiting for clean CI

**Review by JackBoosY — APPROVED** [review]
